{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "20772316",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-22T17:55:35.467711Z",
     "start_time": "2023-11-22T17:55:35.094678Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "cac5ab8e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-22T17:55:35.498363Z",
     "start_time": "2023-11-22T17:55:35.495581Z"
    }
   },
   "outputs": [],
   "source": [
    "# identify number of lines of metadata to skip\n",
    "def metadata_line_count(file_path):\n",
    "    with open(file_path, 'r') as file:\n",
    "        count = 0\n",
    "        find_str = \"#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\"\n",
    "        # should be way less than 10 thousand lines of metadata\n",
    "        while count < 10**4:\n",
    "            line = file.readline().strip()\n",
    "            if find_str in line:\n",
    "                return count\n",
    "            count += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "11716e0d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-22T17:55:36.060430Z",
     "start_time": "2023-11-22T17:55:36.055931Z"
    }
   },
   "outputs": [],
   "source": [
    "def read_vcf(file_path, metadata_line_count):\n",
    "    df = pd.read_csv(file_path, delimiter='\\t', header=0, skiprows=metadata_line_count)\n",
    "    \n",
    "    # Split INFO column\n",
    "    info_split = df['INFO'].str.split(';', expand=True)\n",
    "\n",
    "    # Rename the new columns with 'INFO_' prefix\n",
    "    info_split.columns = [f'INFO_{i}' for i in range(1, info_split.shape[1] + 1)]\n",
    "\n",
    "    # Concatenate the original DataFrame with the new columns\n",
    "    df = pd.concat([df, info_split], axis=1)\n",
    "    \n",
    "    # delete the previous INFO column now that we have all of the split INFO columns\n",
    "    df = df.drop('INFO', axis=1)\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9c57b5da",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-22T17:55:36.992308Z",
     "start_time": "2023-11-22T17:55:36.988730Z"
    }
   },
   "outputs": [],
   "source": [
    "def create_len_col(df, column_name):\n",
    "    \"\"\"\n",
    "    Create a new column containing the lengths of values in a specified column.\n",
    "    Example use case: create a column for the length of alternative and reference reads\n",
    "    \n",
    "    Parameters:\n",
    "    - df (pd.DataFrame): The input DataFrame.\n",
    "    - column_name (str): Name of the column for which length is calculated.\n",
    "\n",
    "    Returns:\n",
    "    - pd.DataFrame: DataFrame with a new column containing lengths of specified values.\n",
    "    \"\"\"\n",
    "    new_col_name = column_name + '_LEN'\n",
    "    # Only add length column if this column name doesn't already exist\n",
    "    if new_col_name not in df.columns:\n",
    "        df[new_col_name] = df[column_name].apply(len)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f6463e0d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-22T17:55:37.734747Z",
     "start_time": "2023-11-22T17:55:37.731179Z"
    }
   },
   "outputs": [],
   "source": [
    "def create_snp_id_col(df, column_list, truncate_columns=None):\n",
    "    \"\"\"\n",
    "    Concatenate list of columns in the given order and create a new column as the SNP ID\n",
    "\n",
    "    Parameters:\n",
    "    - df (pd.DataFrame): The input DataFrame.\n",
    "    - column_list (list): List of column names to concatenate.\n",
    "    - order of the column names matters ie. CHR, POS, REF, ALT = chr1_2131_A_C vs ALT, REF ...C_A\n",
    "    Returns:\n",
    "    - pd.DataFrame: DataFrame with the new concatenated ID column.\n",
    "    \"\"\"\n",
    "    # Ensures all specified columns exist in the DataFrame\n",
    "    for column in column_list:\n",
    "        if column not in df.columns:\n",
    "            raise ValueError(f\"Column '{column}' does not exist in the DataFrame.\")\n",
    "    \n",
    "    # Create a copy of the DataFrame to avoid modifying the original during truncation\n",
    "    df_copy = df.copy()\n",
    "\n",
    "    # Truncate each column value if the column is in truncate columns to a maximum length of 5 characters\n",
    "    for column in column_list:\n",
    "        if truncate_columns is not None and column in truncate_columns:\n",
    "            df_copy[column] = df_copy[column].astype(str).apply(lambda x: x[:5])\n",
    "\n",
    "    \n",
    "    # Concatenate columns in the specified order\n",
    "    df['SNP_ID'] = df_copy[column_list].astype(str).agg('_'.join, axis=1)\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f6886afd",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-22T17:55:38.463863Z",
     "start_time": "2023-11-22T17:55:38.460603Z"
    }
   },
   "outputs": [],
   "source": [
    "def split_select_col(df, column, keyword, new_col_name, multiple_columns, remove_keyword, delimiter='|'):\n",
    "    split_values = df[column].str.split(delimiter, expand=True)\n",
    "    found = False\n",
    "    for col in range(len(split_values.columns)):\n",
    "        if keyword in split_values.iloc[0][col]:\n",
    "            if remove_keyword:\n",
    "                selected_col = split_values.iloc[:, col].str.replace(keyword, '')\n",
    "            else:\n",
    "                selected_col = split_values.iloc[:, col]\n",
    "            selected_col.name = new_col_name\n",
    "            df = pd.concat([df, selected_col], axis=1)\n",
    "            print(col)\n",
    "            found = True\n",
    "            break\n",
    "    if not found:\n",
    "        print('keyword not found in columns')\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "017394f9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-22T17:55:39.137972Z",
     "start_time": "2023-11-22T17:55:39.133288Z"
    }
   },
   "outputs": [],
   "source": [
    "# PL stands for \"Phred-scaled likelihoods.\" which is the column of probabilities associated with\tGT:AD:DP:GQ:PL\n",
    "# for now this only extracts DP and AD values\n",
    "def split_pl(df, column, extract_AD=True, extract_DP=True, extract_alt_read_freq=True):\n",
    "    # need to extract AD and DP columns if we are extracting ALT_READ frequency\n",
    "    if extract_alt_read_freq:\n",
    "        extract_AD, extract_DP = True, True\n",
    "        \n",
    "    if extract_AD:\n",
    "        # Splitting based on ':' delimiter and extracting the 2nd column\n",
    "        # The second column would be AD from 'GT:AD:DP:GQ:PL'\n",
    "        df['AD'] = df[column].apply(lambda x: x.split(':')[1])\n",
    "        # Further splitting the 'AD' column into two columns, left for Alt Reads, right Ref\n",
    "        split_AD = df['AD'].str.split(',', expand=True)\n",
    "        df[['REF_READS', 'ALT_READS']] = split_AD.iloc[:, :2]\n",
    "        # Convert REF_READS, ALT_READS to integer columns\n",
    "        df['REF_READS'], df['ALT_READS'] = df['REF_READS'].astype(int), df['ALT_READS'].astype(int)\n",
    "        # Deleting the AD column\n",
    "        del df['AD']\n",
    "    \n",
    "    if extract_DP:\n",
    "        # Splitting based on ':' delimiter and extracting the 3rd column\n",
    "        # The second column would be DP from 'GT:AD:DP:GQ:PL'\n",
    "        df['DP'] = df[column].apply(lambda x: x.split(':')[2])\n",
    "        df['DP'] = df['DP'].astype(int)\n",
    "        \n",
    "    if extract_alt_read_freq:\n",
    "        # Calculate ALT READ frequency by dividing it by total depth, round to 8 decimal places\n",
    "        df['ALT_FREQ'] = (df['ALT_READS'] / df['DP']).round(8)\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "cde37294",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-22T17:55:43.876040Z",
     "start_time": "2023-11-22T17:55:43.871377Z"
    }
   },
   "outputs": [],
   "source": [
    "def filter_variant_type(df, column, keywords, extract=True, extract_all_keywords=False):\n",
    "    if extract_all_keywords:\n",
    "        extract = True\n",
    "        \n",
    "    for keyword in keywords:\n",
    "        # create a new column that states whether the variant type of the value contains keyword\n",
    "        df[keyword + '_exists'] = df[column].str.contains(keyword, case=False)\n",
    "    if extract:\n",
    "        # sum of boolean values will determine if at least one is True/one keyword present\n",
    "        keyword_columns = [keyword + '_exists' for keyword in keywords]\n",
    "        df['ANY_KEYWORD'] = df[keyword_columns].any(axis=1)\n",
    "        # if no keywords in variant type, remove the row\n",
    "        df = df[df['ANY_KEYWORD']]\n",
    "        del df['ANY_KEYWORD']\n",
    "    # creating a separate column of all the keywords in that row by concatenating columns\n",
    "    # for each column turn it into strings column, if it is True change value to keyword else ''\n",
    "    # then combine all keyword columns, then remove last character which will always be a ','\n",
    "    # initialize this column with empty strings and iteratively add each column as they are made\n",
    "    df['VARIANT_TYPE'] = ''\n",
    "    for keyword in keywords:\n",
    "        keyword_column_name = keyword +'_exists'\n",
    "        df[keyword_column_name] = df[keyword_column_name].replace({True: keyword + ',', False: ''})    \n",
    "        df['VARIANT_TYPE'] = df['VARIANT_TYPE'] + df[keyword_column_name]\n",
    "        del df[keyword_column_name]\n",
    "    # specified in parameter extract_all_keywords, just remove the comma at end of all strings\n",
    "    if extract_all_keywords:\n",
    "        # removal of the last character which will always be a column since we sum keyword+',' columns\n",
    "        df['VARIANT_TYPE'] = df['VARIANT_TYPE'].str.slice(0, -1)\n",
    "    # otherwise, remove everything past the first index of comma, leaving only first value\n",
    "    else:\n",
    "        # get first index of comma\n",
    "        df['COMMA_IDX'] = df['VARIANT_TYPE'].str.find(',').astype('int')\n",
    "        # remove rest of string after comma, inclusive of comma, i.e. hello, hi --> hello\n",
    "        df['VARIANT_TYPE'] = df.apply(lambda row: row['VARIANT_TYPE'][:row['COMMA_IDX']], axis=1)\n",
    "        del df['COMMA_IDX']\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "e49c6114",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-18T04:58:19.936884Z",
     "start_time": "2023-11-18T04:58:19.933085Z"
    }
   },
   "outputs": [],
   "source": [
    "def extract_AA_change(df, column, delimeter='|', truncate=True):\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "e09c49bd",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-18T05:25:11.053026Z",
     "start_time": "2023-11-18T05:25:10.858169Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n",
      "0.9956980836918263\n"
     ]
    }
   ],
   "source": [
    "split_values = df['INFO_19'].str.split('|', expand=True).iloc[:, 4]\n",
    "print(split_values.str.contains('gene-').all())\n",
    "print(split_values.str.count('gene-').sum() / len(split_values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4cbb5ef4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-22T18:00:55.335270Z",
     "start_time": "2023-11-22T18:00:53.279414Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4\n",
      "10\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_139023/1273072395.py:19: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['VARIANT_TYPE'] = ''\n",
      "/tmp/ipykernel_139023/1273072395.py:22: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df[keyword_column_name] = df[keyword_column_name].replace({True: keyword + ',', False: ''})\n",
      "/tmp/ipykernel_139023/1273072395.py:23: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['VARIANT_TYPE'] = df['VARIANT_TYPE'] + df[keyword_column_name]\n",
      "/tmp/ipykernel_139023/1273072395.py:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['COMMA_IDX'] = df['VARIANT_TYPE'].str.find(',').astype('int')\n",
      "/tmp/ipykernel_139023/1273072395.py:34: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['VARIANT_TYPE'] = df.apply(lambda row: row['VARIANT_TYPE'][:row['COMMA_IDX']], axis=1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>100M1</th>\n",
       "      <th>INFO_1</th>\n",
       "      <th>...</th>\n",
       "      <th>REF_LEN</th>\n",
       "      <th>ALT_LEN</th>\n",
       "      <th>SNP_ID</th>\n",
       "      <th>Gene_Symbol</th>\n",
       "      <th>REF_READS</th>\n",
       "      <th>ALT_READS</th>\n",
       "      <th>DP</th>\n",
       "      <th>ALT_FREQ</th>\n",
       "      <th>VARIANT_TYPE</th>\n",
       "      <th>Protein</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3740866</td>\n",
       "      <td>.</td>\n",
       "      <td>TC</td>\n",
       "      <td>T</td>\n",
       "      <td>3487.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:10,104:114:30:3516,0,30</td>\n",
       "      <td>AC=1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1_3740866_TC_T_2_1</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>10</td>\n",
       "      <td>104</td>\n",
       "      <td>114</td>\n",
       "      <td>0.912281</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>p.Arg235fs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6872919</td>\n",
       "      <td>.</td>\n",
       "      <td>GC</td>\n",
       "      <td>G</td>\n",
       "      <td>4570.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>1/1:8,142:150:99:4599,183,0</td>\n",
       "      <td>AC=2</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1_6872919_GC_G_2_1</td>\n",
       "      <td>St18</td>\n",
       "      <td>8</td>\n",
       "      <td>142</td>\n",
       "      <td>150</td>\n",
       "      <td>0.946667</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>p.Pro226fs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>8677327</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>4131.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>1/1:19,171:190:47:4160,47,0</td>\n",
       "      <td>AC=2</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1_8677327_A_T_1_1</td>\n",
       "      <td>Sntg1</td>\n",
       "      <td>19</td>\n",
       "      <td>171</td>\n",
       "      <td>190</td>\n",
       "      <td>0.900000</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>p.Ser172Arg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9615822</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>334.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>GT:AD:DP:GQ:PGT:PID:PL</td>\n",
       "      <td>0/1:33,11:44:99:0|1:9615802_G_A:363,0,2497</td>\n",
       "      <td>AC=1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1_9615822_C_T_1_1</td>\n",
       "      <td>Rrs1</td>\n",
       "      <td>33</td>\n",
       "      <td>11</td>\n",
       "      <td>44</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>p.Thr25Met</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9615846</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>262.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>GT:AD:DP:GQ:PGT:PID:PL</td>\n",
       "      <td>0/1:36,10:46:99:0|1:9615802_G_A:291,0,2489</td>\n",
       "      <td>AC=1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1_9615846_A_T_1_1</td>\n",
       "      <td>Rrs1</td>\n",
       "      <td>36</td>\n",
       "      <td>10</td>\n",
       "      <td>46</td>\n",
       "      <td>0.217391</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>p.Glu33Val</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10322</th>\n",
       "      <td>chrX</td>\n",
       "      <td>162979973</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>1849.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:33,74:107:99:1878,0,606</td>\n",
       "      <td>AC=1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>chrX_162979973_T_A_1_1</td>\n",
       "      <td>Bmx</td>\n",
       "      <td>33</td>\n",
       "      <td>74</td>\n",
       "      <td>107</td>\n",
       "      <td>0.691589</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>p.Lys604Met</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10325</th>\n",
       "      <td>chrX</td>\n",
       "      <td>163739523</td>\n",
       "      <td>.</td>\n",
       "      <td>CA</td>\n",
       "      <td>C</td>\n",
       "      <td>998.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:89,58:160:99:1027,0,1777</td>\n",
       "      <td>AC=1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>chrX_163739523_CA_C_2_1</td>\n",
       "      <td>Mospd2</td>\n",
       "      <td>89</td>\n",
       "      <td>58</td>\n",
       "      <td>160</td>\n",
       "      <td>0.362500</td>\n",
       "      <td>splice_acceptor</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10338</th>\n",
       "      <td>chrX</td>\n",
       "      <td>166090512</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>1729.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:97,73:170:99:1758,0,2595</td>\n",
       "      <td>AC=1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>chrX_166090512_A_T_1_1</td>\n",
       "      <td>Tlr7</td>\n",
       "      <td>97</td>\n",
       "      <td>73</td>\n",
       "      <td>170</td>\n",
       "      <td>0.429412</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>p.Tyr328Asn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10339</th>\n",
       "      <td>chrX</td>\n",
       "      <td>166260331</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>2804.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:75,103:178:99:2833,0,1717</td>\n",
       "      <td>AC=1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>chrX_166260331_T_A_1_1</td>\n",
       "      <td>Frmpd4</td>\n",
       "      <td>75</td>\n",
       "      <td>103</td>\n",
       "      <td>178</td>\n",
       "      <td>0.578652</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>p.Asp1155Val</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10340</th>\n",
       "      <td>chrX</td>\n",
       "      <td>167579241</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>604.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:40,29:69:99:633,0,1024</td>\n",
       "      <td>AC=1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>chrX_167579241_A_T_1_1</td>\n",
       "      <td>Arhgap6</td>\n",
       "      <td>40</td>\n",
       "      <td>29</td>\n",
       "      <td>69</td>\n",
       "      <td>0.420290</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>p.Ser9Cys</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2557 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      #CHROM        POS ID REF ALT     QUAL FILTER                  FORMAT  \\\n",
       "0       chr1    3740866  .  TC   T  3487.77   PASS          GT:AD:DP:GQ:PL   \n",
       "3       chr1    6872919  .  GC   G  4570.77   PASS          GT:AD:DP:GQ:PL   \n",
       "7       chr1    8677327  .   A   T  4131.77   PASS          GT:AD:DP:GQ:PL   \n",
       "10      chr1    9615822  .   C   T   334.77   PASS  GT:AD:DP:GQ:PGT:PID:PL   \n",
       "11      chr1    9615846  .   A   T   262.77   PASS  GT:AD:DP:GQ:PGT:PID:PL   \n",
       "...      ...        ... ..  ..  ..      ...    ...                     ...   \n",
       "10322   chrX  162979973  .   T   A  1849.77   PASS          GT:AD:DP:GQ:PL   \n",
       "10325   chrX  163739523  .  CA   C   998.77   PASS          GT:AD:DP:GQ:PL   \n",
       "10338   chrX  166090512  .   A   T  1729.77   PASS          GT:AD:DP:GQ:PL   \n",
       "10339   chrX  166260331  .   T   A  2804.77   PASS          GT:AD:DP:GQ:PL   \n",
       "10340   chrX  167579241  .   A   T   604.77   PASS          GT:AD:DP:GQ:PL   \n",
       "\n",
       "                                            100M1 INFO_1  ... REF_LEN ALT_LEN  \\\n",
       "0                     0/1:10,104:114:30:3516,0,30   AC=1  ...       2       1   \n",
       "3                     1/1:8,142:150:99:4599,183,0   AC=2  ...       2       1   \n",
       "7                     1/1:19,171:190:47:4160,47,0   AC=2  ...       1       1   \n",
       "10     0/1:33,11:44:99:0|1:9615802_G_A:363,0,2497   AC=1  ...       1       1   \n",
       "11     0/1:36,10:46:99:0|1:9615802_G_A:291,0,2489   AC=1  ...       1       1   \n",
       "...                                           ...    ...  ...     ...     ...   \n",
       "10322                 0/1:33,74:107:99:1878,0,606   AC=1  ...       1       1   \n",
       "10325                0/1:89,58:160:99:1027,0,1777   AC=1  ...       2       1   \n",
       "10338                0/1:97,73:170:99:1758,0,2595   AC=1  ...       1       1   \n",
       "10339               0/1:75,103:178:99:2833,0,1717   AC=1  ...       1       1   \n",
       "10340                  0/1:40,29:69:99:633,0,1024   AC=1  ...       1       1   \n",
       "\n",
       "                        SNP_ID Gene_Symbol REF_READS ALT_READS   DP  ALT_FREQ  \\\n",
       "0        chr1_3740866_TC_T_2_1        Xkr4        10       104  114  0.912281   \n",
       "3        chr1_6872919_GC_G_2_1        St18         8       142  150  0.946667   \n",
       "7         chr1_8677327_A_T_1_1       Sntg1        19       171  190  0.900000   \n",
       "10        chr1_9615822_C_T_1_1        Rrs1        33        11   44  0.250000   \n",
       "11        chr1_9615846_A_T_1_1        Rrs1        36        10   46  0.217391   \n",
       "...                        ...         ...       ...       ...  ...       ...   \n",
       "10322   chrX_162979973_T_A_1_1         Bmx        33        74  107  0.691589   \n",
       "10325  chrX_163739523_CA_C_2_1      Mospd2        89        58  160  0.362500   \n",
       "10338   chrX_166090512_A_T_1_1        Tlr7        97        73  170  0.429412   \n",
       "10339   chrX_166260331_T_A_1_1      Frmpd4        75       103  178  0.578652   \n",
       "10340   chrX_167579241_A_T_1_1     Arhgap6        40        29   69  0.420290   \n",
       "\n",
       "             VARIANT_TYPE       Protein  \n",
       "0      frameshift_variant    p.Arg235fs  \n",
       "3      frameshift_variant    p.Pro226fs  \n",
       "7        missense_variant   p.Ser172Arg  \n",
       "10       missense_variant    p.Thr25Met  \n",
       "11       missense_variant    p.Glu33Val  \n",
       "...                   ...           ...  \n",
       "10322    missense_variant   p.Lys604Met  \n",
       "10325     splice_acceptor                \n",
       "10338    missense_variant   p.Tyr328Asn  \n",
       "10339    missense_variant  p.Asp1155Val  \n",
       "10340    missense_variant     p.Ser9Cys  \n",
       "\n",
       "[2557 rows x 40 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# EXAMPLE TEST\n",
    "file_path = '100M1.final.vcf'\n",
    "metadata_lines = metadata_line_count(file_path)\n",
    "df = read_vcf(file_path, metadata_lines)\n",
    "df = create_len_col(df, 'REF')\n",
    "df = create_len_col(df, 'ALT')\n",
    "df = create_snp_id_col(df, ['#CHROM', 'POS', 'REF', 'ALT', 'REF_LEN', 'ALT_LEN'], ['REF', 'ALT'])\n",
    "df = split_select_col(df, 'INFO_19', 'gene-', 'Gene_Symbol', multiple_columns=False, remove_keyword=True)\n",
    "df = split_pl(df, '100M1')\n",
    "filter_variant_keywords = ['disruptive_inframe_deletion', 'disruptive_inframe_insertion', 'frameshift_variant', 'stop_lost', 'stop_gained', 'missense_variant', 'splice_acceptor', 'splice_donor', 'inframe_deletion']\n",
    "df = filter_variant_type(df, column='INFO_19', keywords=filter_variant_keywords)\n",
    "df = split_select_col(df, 'INFO_19', 'p.', 'Protein', multiple_columns=True, remove_keyword=False)\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "86dd7f8f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-10T23:30:43.340370Z",
     "start_time": "2023-11-10T23:30:43.338089Z"
    }
   },
   "outputs": [],
   "source": [
    "line2 = \"\"\"#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\t100M1\"\"\"\n",
    "line = \"\"\"chr1\t3740866\t.\tTC\tT\t3487.77\tPASS\tAC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-3.460e-01;ClippingRankSum=0.426;DP=116;FS=0.000;MLEAC=1;MLEAF=0.500;MQ=60.19;MQRankSum=1.32;QD=30.59;ReadPosRankSum=0.175;SOR=0.625;set=variant2;DEL;HET;VARTYPE=DEL;ANN=T|frameshift_variant|HIGH|Xkr4|gene-Xkr4|transcript|rna-NM_001011874.1|Coding|1/3|c.705delG|p.Arg235fs|855/3634|705/1944|235/647||,T|frameshift_variant|HIGH|Xkr4|gene-Xkr4|transcript|rna-XM_006495550.5|Coding|1/4|c.705delG|p.Arg235fs|867/12274|705/1944|235/647||,T|frameshift_variant|HIGH|Xkr4|gene-Xkr4|transcript|rna-XM_011238395.4|Coding|1/3|c.705delG|p.Arg235fs|867/2104|705/1011|235/336||,T|frameshift_variant|HIGH|Xkr4|gene-Xkr4|transcript|rna-XM_036152287.1|Coding|1/3|c.705delG|p.Arg235fs|867/2764|705/1026|235/341||,T|downstream_gene_variant|MODIFIER|Gene_gene-LOC108167595|Gene_gene-LOC108167595|transcript|gene-LOC108167595|Noncoding||n.*969delG|||||866|;LOF=(Xkr4|gene-Xkr4|4|1.00)\tGT:AD:DP:GQ:PL\t0/1:10,104:114:30:3516,0,30\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "43fb49fb",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-10T23:30:44.078828Z",
     "start_time": "2023-11-10T23:30:44.075984Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['chr1', '3740866', '.', 'TC', 'T', '3487.77', 'PASS', 'AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-3.460e-01;ClippingRankSum=0.426;DP=116;FS=0.000;MLEAC=1;MLEAF=0.500;MQ=60.19;MQRankSum=1.32;QD=30.59;ReadPosRankSum=0.175;SOR=0.625;set=variant2;DEL;HET;VARTYPE=DEL;ANN=T|frameshift_variant|HIGH|Xkr4|gene-Xkr4|transcript|rna-NM_001011874.1|Coding|1/3|c.705delG|p.Arg235fs|855/3634|705/1944|235/647||,T|frameshift_variant|HIGH|Xkr4|gene-Xkr4|transcript|rna-XM_006495550.5|Coding|1/4|c.705delG|p.Arg235fs|867/12274|705/1944|235/647||,T|frameshift_variant|HIGH|Xkr4|gene-Xkr4|transcript|rna-XM_011238395.4|Coding|1/3|c.705delG|p.Arg235fs|867/2104|705/1011|235/336||,T|frameshift_variant|HIGH|Xkr4|gene-Xkr4|transcript|rna-XM_036152287.1|Coding|1/3|c.705delG|p.Arg235fs|867/2764|705/1026|235/341||,T|downstream_gene_variant|MODIFIER|Gene_gene-LOC108167595|Gene_gene-LOC108167595|transcript|gene-LOC108167595|Noncoding||n.*969delG|||||866|;LOF=(Xkr4|gene-Xkr4|4|1.00)', 'GT:AD:DP:GQ:PL', '0/1:10,104:114:30:3516,0,30']\n"
     ]
    }
   ],
   "source": [
    "print(line.split())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a5102a0c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-10T23:24:31.238613Z",
     "start_time": "2023-11-10T23:24:27.120438Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Chr\tStart\tEnd\tRef\tAlt\tabraom_freq\tabraom_filter\tabraom_cegh_filter\tALL.sites.2015_08\tavsnp150\tCLNALLELEID\tCLNDN\tCLNDISDB\tCLNREVSTAT\tCLNSIG\tcosmic70\tStart_hg19\tDamagePredCount\tSIFT_pred\tSIFT4G_pred\tLRT_pred\tMutationTaster_pred\tMutationAssessor_pred\tFATHMM_pred\tPROVEAN_pred\tMetaSVM_pred\tMetaLR_pred\tM-CAP_pred\tMutPred_score\tMVP_score\tMPC_score\tPrimateAI_pred\tDEOGEN2_pred\tBayesDel_addAF_pred\tBayesDel_noAF_pred\tLIST-S2_pred\tDANN_score\tfathmm-MKL_coding_pred\tfathmm-XF_coding_pred\tEigen-raw_coding\tEigen-phred_coding\tEigen-PC-raw_coding\tEigen-PC-phred_coding\tintegrated_fitCons_score\tGM12878_fitCons_score\tH1-hESC_fitCons_score\tHUVEC_fitCons_score\tGERP++_NR\tGERP++_RS\tphyloP100way_vertebrate\tphyloP30way_mammalian\tphyloP17way_primate\tphastCons100way_vertebrate\tphastCons30way_mammalian\tphastCons17way_primate\tbStatistic\tInterpro_domain\tGTEx_V8_gene\tGTEx_V8_tissue\tesp6500siv2_all\tExAC_ALL\tExAC_AFR\tExAC_AMR\tExAC_EAS\tExAC_FIN\tExAC_NFE\tExAC_OTH\tExAC_SAS\tDN ID\tPatient ID\tPhenotype\tPlatform\tStudy\tPubmed ID\tGME_AF\tGME_NWA\tGME_NEA\tGME_AP\tGME_Israel\tGME_SD\tGME_TP\tGME_CA\tAF\tAF_popmax\tAF_male\tAF_female\tAF_raw\tAF_afr\tAF_sas\tAF_amr\tAF_eas\tAF_nfe\tAF_fin\tAF_asj\tAF_oth\tnon_topmed_AF_popmax\tnon_neuro_AF_popmax\tnon_cancer_AF_popmax\tcontrols_AF_popmax\tHRC_AF\tHRC_AC\tHRC_AN\tHRC_non1000G_AF\tHRC_non1000G_AC\tHRC_non1000G_AN\tICGC_Id\tICGC_Occurrence\tInterVar_automated\tPVS1\tPS1\tPS2\tPS3\tPS4\tPM1\tPM2\tPM3\tPM4\tPM5\tPM6\tPP1\tPP2\tPP3\tPP4\tPP5\tBA1\tBS1\tBS2\tBS3\tBS4\tBP1\tBP2\tBP3\tBP4\tBP5\tBP6\tBP7\tKaviar_AF\tKaviar_AC\tKaviar_AN\tMCAP\tREVEL\tOtherinfo1\tOtherinfo2\tOtherinfo3\tOtherinfo4\tOtherinfo5\tOtherinfo6\tOtherinfo7\tOtherinfo8\tOtherinfo9\tOtherinfo10\tOtherinfo11\tOtherinfo12\tOtherinfo13\n",
      "a\n"
     ]
    }
   ],
   "source": [
    "file_path = '222511.hg38_multianno.txt'\n",
    "\n",
    "with open(file_path, 'r') as file:\n",
    "    while True:\n",
    "        x = input()\n",
    "        if x == \"\":\n",
    "            print(file.readline().strip())\n",
    "        else:\n",
    "            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "98826952",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-10T23:43:25.292735Z",
     "start_time": "2023-11-10T23:43:25.191560Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                       0\n",
      "0      #CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\...\n",
      "1      chr1\\t3740866\\t.\\tTC\\tT\\t3487.77\\tPASS\\tAC=1;S...\n",
      "2      chr1\\t6304093\\t.\\tTTG\\tT\\t31.77\\tMG_INDEL_Filt...\n",
      "3      chr1\\t6841143\\t.\\tCT\\tC\\t75.77\\tMG_INDEL_Filte...\n",
      "4      chr1\\t6872919\\t.\\tGC\\tG\\t4570.77\\tPASS\\tAC=2;S...\n",
      "...                                                  ...\n",
      "10339  chrX\\t166090512\\t.\\tA\\tT\\t1729.77\\tPASS\\tAC=1;...\n",
      "10340  chrX\\t166260331\\t.\\tT\\tA\\t2804.77\\tPASS\\tAC=1;...\n",
      "10341  chrX\\t167579241\\t.\\tA\\tT\\t604.77\\tPASS\\tAC=1;S...\n",
      "10342  chrX\\t168084638\\t.\\tT\\tC\\t218.77\\tPASS\\tAC=1;S...\n",
      "10343  chrX\\t168559972\\t.\\tG\\tT\\t1803.77\\tPASS\\tAC=1;...\n",
      "\n",
      "[10344 rows x 1 columns]\n",
      "150\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_373592/1067633238.py:23: FutureWarning: In a future version of pandas all arguments of read_csv except for the argument 'filepath_or_buffer' will be keyword-only.\n",
      "  df = pd.read_csv(file_path, delimiter, header=None, skiprows=skiprows-6)\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Replace 'your_file.txt' with the actual path to your file\n",
    "file_path = '100M1.final.vcf'\n",
    "\n",
    "# Specify the delimiter (space in this case)\n",
    "delimiter = ' '\n",
    "\n",
    "# Number of lines to skip at the beginning\n",
    "skiprows = 0\n",
    "\n",
    "# Open the file and read lines until a line with the correct delimiter is found\n",
    "with open(file_path, 'r') as file:\n",
    "    while True:\n",
    "        line = file.readline()\n",
    "        skiprows += 1\n",
    "\n",
    "        # Check if the line contains the expected delimiter\n",
    "        if 'GT:AD:DP:GQ:PL' in line:\n",
    "            break\n",
    "\n",
    "# Read the text file into a DataFrame, skipping the lines until the delimiter is found\n",
    "df = pd.read_csv(file_path, delimiter, header=None, skiprows=skiprows-6)\n",
    "\n",
    "# Display the DataFrame\n",
    "print(df)\n",
    "print(skiprows)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "aff884f8",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-10T23:43:27.322838Z",
     "start_time": "2023-11-10T23:43:27.297339Z"
    }
   },
   "outputs": [],
   "source": [
    "df = df[0].str.split('\\t', expand=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "2043ee2f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-10T23:43:27.702836Z",
     "start_time": "2023-11-10T23:43:27.692715Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>#CHROM</td>\n",
       "      <td>POS</td>\n",
       "      <td>ID</td>\n",
       "      <td>REF</td>\n",
       "      <td>ALT</td>\n",
       "      <td>QUAL</td>\n",
       "      <td>FILTER</td>\n",
       "      <td>INFO</td>\n",
       "      <td>FORMAT</td>\n",
       "      <td>100M1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3740866</td>\n",
       "      <td>.</td>\n",
       "      <td>TC</td>\n",
       "      <td>T</td>\n",
       "      <td>3487.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-3.460...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:10,104:114:30:3516,0,30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6304093</td>\n",
       "      <td>.</td>\n",
       "      <td>TTG</td>\n",
       "      <td>T</td>\n",
       "      <td>31.77</td>\n",
       "      <td>MG_INDEL_Filter</td>\n",
       "      <td>AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-1.391...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:30,5:35:60:60,0,838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6841143</td>\n",
       "      <td>.</td>\n",
       "      <td>CT</td>\n",
       "      <td>C</td>\n",
       "      <td>75.77</td>\n",
       "      <td>MG_INDEL_Filter</td>\n",
       "      <td>AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=0.727;...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:45,11:58:99:104,0,830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6872919</td>\n",
       "      <td>.</td>\n",
       "      <td>GC</td>\n",
       "      <td>G</td>\n",
       "      <td>4570.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=2;SAMPLES_AF=1.00;AN=2;BaseQRankSum=0.130;C...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>1/1:8,142:150:99:4599,183,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10339</th>\n",
       "      <td>chrX</td>\n",
       "      <td>166090512</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>1729.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-5.350...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:97,73:170:99:1758,0,2595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10340</th>\n",
       "      <td>chrX</td>\n",
       "      <td>166260331</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>2804.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=1.91;C...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:75,103:178:99:2833,0,1717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10341</th>\n",
       "      <td>chrX</td>\n",
       "      <td>167579241</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>604.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-1.100...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:40,29:69:99:633,0,1024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10342</th>\n",
       "      <td>chrX</td>\n",
       "      <td>168084638</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>218.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=3.57;C...</td>\n",
       "      <td>GT:AD:DP:GQ:PGT:PID:PL</td>\n",
       "      <td>0/1:13,7:20:99:0|1:168084632_G_T:247,0,342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10343</th>\n",
       "      <td>chrX</td>\n",
       "      <td>168559972</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1803.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-1.983...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:158,86:244:99:1832,0,4168</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10344 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            0          1   2    3    4        5                6  \\\n",
       "0      #CHROM        POS  ID  REF  ALT     QUAL           FILTER   \n",
       "1        chr1    3740866   .   TC    T  3487.77             PASS   \n",
       "2        chr1    6304093   .  TTG    T    31.77  MG_INDEL_Filter   \n",
       "3        chr1    6841143   .   CT    C    75.77  MG_INDEL_Filter   \n",
       "4        chr1    6872919   .   GC    G  4570.77             PASS   \n",
       "...       ...        ...  ..  ...  ...      ...              ...   \n",
       "10339    chrX  166090512   .    A    T  1729.77             PASS   \n",
       "10340    chrX  166260331   .    T    A  2804.77             PASS   \n",
       "10341    chrX  167579241   .    A    T   604.77             PASS   \n",
       "10342    chrX  168084638   .    T    C   218.77             PASS   \n",
       "10343    chrX  168559972   .    G    T  1803.77             PASS   \n",
       "\n",
       "                                                       7  \\\n",
       "0                                                   INFO   \n",
       "1      AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-3.460...   \n",
       "2      AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-1.391...   \n",
       "3      AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=0.727;...   \n",
       "4      AC=2;SAMPLES_AF=1.00;AN=2;BaseQRankSum=0.130;C...   \n",
       "...                                                  ...   \n",
       "10339  AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-5.350...   \n",
       "10340  AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=1.91;C...   \n",
       "10341  AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-1.100...   \n",
       "10342  AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=3.57;C...   \n",
       "10343  AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-1.983...   \n",
       "\n",
       "                            8                                           9  \n",
       "0                      FORMAT                                       100M1  \n",
       "1              GT:AD:DP:GQ:PL                 0/1:10,104:114:30:3516,0,30  \n",
       "2              GT:AD:DP:GQ:PL                     0/1:30,5:35:60:60,0,838  \n",
       "3              GT:AD:DP:GQ:PL                   0/1:45,11:58:99:104,0,830  \n",
       "4              GT:AD:DP:GQ:PL                 1/1:8,142:150:99:4599,183,0  \n",
       "...                       ...                                         ...  \n",
       "10339          GT:AD:DP:GQ:PL                0/1:97,73:170:99:1758,0,2595  \n",
       "10340          GT:AD:DP:GQ:PL               0/1:75,103:178:99:2833,0,1717  \n",
       "10341          GT:AD:DP:GQ:PL                  0/1:40,29:69:99:633,0,1024  \n",
       "10342  GT:AD:DP:GQ:PGT:PID:PL  0/1:13,7:20:99:0|1:168084632_G_T:247,0,342  \n",
       "10343          GT:AD:DP:GQ:PL               0/1:158,86:244:99:1832,0,4168  \n",
       "\n",
       "[10344 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "a975f3b2",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-10T23:43:58.308034Z",
     "start_time": "2023-11-10T23:43:58.299241Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>100M1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3740866</td>\n",
       "      <td>.</td>\n",
       "      <td>TC</td>\n",
       "      <td>T</td>\n",
       "      <td>3487.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-3.460...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:10,104:114:30:3516,0,30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6304093</td>\n",
       "      <td>.</td>\n",
       "      <td>TTG</td>\n",
       "      <td>T</td>\n",
       "      <td>31.77</td>\n",
       "      <td>MG_INDEL_Filter</td>\n",
       "      <td>AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-1.391...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:30,5:35:60:60,0,838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6841143</td>\n",
       "      <td>.</td>\n",
       "      <td>CT</td>\n",
       "      <td>C</td>\n",
       "      <td>75.77</td>\n",
       "      <td>MG_INDEL_Filter</td>\n",
       "      <td>AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=0.727;...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:45,11:58:99:104,0,830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6872919</td>\n",
       "      <td>.</td>\n",
       "      <td>GC</td>\n",
       "      <td>G</td>\n",
       "      <td>4570.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=2;SAMPLES_AF=1.00;AN=2;BaseQRankSum=0.130;C...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>1/1:8,142:150:99:4599,183,0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6878048</td>\n",
       "      <td>.</td>\n",
       "      <td>AT</td>\n",
       "      <td>A</td>\n",
       "      <td>53.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-3.400...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:11,5:16:82:82,0,224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10338</th>\n",
       "      <td>chrX</td>\n",
       "      <td>166090512</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>1729.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-5.350...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:97,73:170:99:1758,0,2595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10339</th>\n",
       "      <td>chrX</td>\n",
       "      <td>166260331</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>2804.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=1.91;C...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:75,103:178:99:2833,0,1717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10340</th>\n",
       "      <td>chrX</td>\n",
       "      <td>167579241</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>604.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-1.100...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:40,29:69:99:633,0,1024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10341</th>\n",
       "      <td>chrX</td>\n",
       "      <td>168084638</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>218.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=3.57;C...</td>\n",
       "      <td>GT:AD:DP:GQ:PGT:PID:PL</td>\n",
       "      <td>0/1:13,7:20:99:0|1:168084632_G_T:247,0,342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10342</th>\n",
       "      <td>chrX</td>\n",
       "      <td>168559972</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1803.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-1.983...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:158,86:244:99:1832,0,4168</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10343 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0     #CHROM        POS ID  REF ALT     QUAL           FILTER  \\\n",
       "0       chr1    3740866  .   TC   T  3487.77             PASS   \n",
       "1       chr1    6304093  .  TTG   T    31.77  MG_INDEL_Filter   \n",
       "2       chr1    6841143  .   CT   C    75.77  MG_INDEL_Filter   \n",
       "3       chr1    6872919  .   GC   G  4570.77             PASS   \n",
       "4       chr1    6878048  .   AT   A    53.77             PASS   \n",
       "...      ...        ... ..  ...  ..      ...              ...   \n",
       "10338   chrX  166090512  .    A   T  1729.77             PASS   \n",
       "10339   chrX  166260331  .    T   A  2804.77             PASS   \n",
       "10340   chrX  167579241  .    A   T   604.77             PASS   \n",
       "10341   chrX  168084638  .    T   C   218.77             PASS   \n",
       "10342   chrX  168559972  .    G   T  1803.77             PASS   \n",
       "\n",
       "0                                                   INFO  \\\n",
       "0      AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-3.460...   \n",
       "1      AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-1.391...   \n",
       "2      AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=0.727;...   \n",
       "3      AC=2;SAMPLES_AF=1.00;AN=2;BaseQRankSum=0.130;C...   \n",
       "4      AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-3.400...   \n",
       "...                                                  ...   \n",
       "10338  AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-5.350...   \n",
       "10339  AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=1.91;C...   \n",
       "10340  AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-1.100...   \n",
       "10341  AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=3.57;C...   \n",
       "10342  AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-1.983...   \n",
       "\n",
       "0                      FORMAT                                       100M1  \n",
       "0              GT:AD:DP:GQ:PL                 0/1:10,104:114:30:3516,0,30  \n",
       "1              GT:AD:DP:GQ:PL                     0/1:30,5:35:60:60,0,838  \n",
       "2              GT:AD:DP:GQ:PL                   0/1:45,11:58:99:104,0,830  \n",
       "3              GT:AD:DP:GQ:PL                 1/1:8,142:150:99:4599,183,0  \n",
       "4              GT:AD:DP:GQ:PL                     0/1:11,5:16:82:82,0,224  \n",
       "...                       ...                                         ...  \n",
       "10338          GT:AD:DP:GQ:PL                0/1:97,73:170:99:1758,0,2595  \n",
       "10339          GT:AD:DP:GQ:PL               0/1:75,103:178:99:2833,0,1717  \n",
       "10340          GT:AD:DP:GQ:PL                  0/1:40,29:69:99:633,0,1024  \n",
       "10341  GT:AD:DP:GQ:PGT:PID:PL  0/1:13,7:20:99:0|1:168084632_G_T:247,0,342  \n",
       "10342          GT:AD:DP:GQ:PL               0/1:158,86:244:99:1832,0,4168  \n",
       "\n",
       "[10343 rows x 10 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.to_csv('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "37636a55",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-10T23:43:46.165623Z",
     "start_time": "2023-11-10T23:43:46.162942Z"
    }
   },
   "outputs": [],
   "source": [
    "df.columns = df.iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "36eb746e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-10T23:54:53.334079Z",
     "start_time": "2023-11-10T23:54:53.142875Z"
    }
   },
   "outputs": [],
   "source": [
    "df.to_csv('delimited_100M1_final.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "28d8c9b9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-10T23:43:54.366719Z",
     "start_time": "2023-11-10T23:43:54.361171Z"
    }
   },
   "outputs": [],
   "source": [
    "df = df.drop(0)\n",
    "df = df.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "e28de266",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-10T23:39:06.124345Z",
     "start_time": "2023-11-10T23:39:06.112860Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The file has 10492 lines.\n"
     ]
    }
   ],
   "source": [
    "file_path = '100M1.final.vcf'  # Replace with the actual path to your file\n",
    "\n",
    "# Open the file and count the lines\n",
    "with open(file_path, 'r') as file:\n",
    "    line_count = sum(1 for line in file)\n",
    "\n",
    "print(f'The file has {line_count} lines.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "bc778ce7",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-10T23:41:24.799245Z",
     "start_time": "2023-11-10T23:41:24.620544Z"
    }
   },
   "outputs": [],
   "source": [
    "df.to_csv('test.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "573fc211",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-10T23:53:04.661735Z",
     "start_time": "2023-11-10T23:53:04.659111Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-3.460e-01;ClippingRankSum=0.426;DP=116;FS=0.000;MLEAC=1;MLEAF=0.500;MQ=60.19;MQRankSum=1.32;QD=30.59;ReadPosRankSum=0.175;SOR=0.625;set=variant2;DEL;HET;VARTYPE=DEL;ANN=T|frameshift_variant|HIGH|Xkr4|gene-Xkr4|transcript|rna-NM_001011874.1|Coding|1/3|c.705delG|p.Arg235fs|855/3634|705/1944|235/647||,T|frameshift_variant|HIGH|Xkr4|gene-Xkr4|transcript|rna-XM_006495550.5|Coding|1/4|c.705delG|p.Arg235fs|867/12274|705/1944|235/647||,T|frameshift_variant|HIGH|Xkr4|gene-Xkr4|transcript|rna-XM_011238395.4|Coding|1/3|c.705delG|p.Arg235fs|867/2104|705/1011|235/336||,T|frameshift_variant|HIGH|Xkr4|gene-Xkr4|transcript|rna-XM_036152287.1|Coding|1/3|c.705delG|p.Arg235fs|867/2764|705/1026|235/341||,T|downstream_gene_variant|MODIFIER|Gene_gene-LOC108167595|Gene_gene-LOC108167595|transcript|gene-LOC108167595|Noncoding||n.*969delG|||||866|;LOF=(Xkr4|gene-Xkr4|4|1.00)\n"
     ]
    }
   ],
   "source": [
    "print(df['INFO'].iloc[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "4eeb8967",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-10T23:53:58.596804Z",
     "start_time": "2023-11-10T23:53:58.537300Z"
    }
   },
   "outputs": [],
   "source": [
    "info_split = df['INFO'].str.split(';', expand=True)\n",
    "\n",
    "# Rename the new columns with 'INFO_' prefix\n",
    "info_split.columns = [f'INFO_{i}' for i in range(1, info_split.shape[1] + 1)]\n",
    "\n",
    "# Concatenate the original DataFrame with the new columns\n",
    "df = pd.concat([df, info_split], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "1f8ff5cb",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-10T23:54:00.417005Z",
     "start_time": "2023-11-10T23:54:00.381520Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>100M1</th>\n",
       "      <th>...</th>\n",
       "      <th>INFO_12</th>\n",
       "      <th>INFO_13</th>\n",
       "      <th>INFO_14</th>\n",
       "      <th>INFO_15</th>\n",
       "      <th>INFO_16</th>\n",
       "      <th>INFO_17</th>\n",
       "      <th>INFO_18</th>\n",
       "      <th>INFO_19</th>\n",
       "      <th>INFO_20</th>\n",
       "      <th>INFO_21</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3740866</td>\n",
       "      <td>.</td>\n",
       "      <td>TC</td>\n",
       "      <td>T</td>\n",
       "      <td>3487.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-3.460...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:10,104:114:30:3516,0,30</td>\n",
       "      <td>...</td>\n",
       "      <td>QD=30.59</td>\n",
       "      <td>ReadPosRankSum=0.175</td>\n",
       "      <td>SOR=0.625</td>\n",
       "      <td>set=variant2</td>\n",
       "      <td>DEL</td>\n",
       "      <td>HET</td>\n",
       "      <td>VARTYPE=DEL</td>\n",
       "      <td>ANN=T|frameshift_variant|HIGH|Xkr4|gene-Xkr4|t...</td>\n",
       "      <td>LOF=(Xkr4|gene-Xkr4|4|1.00)</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6304093</td>\n",
       "      <td>.</td>\n",
       "      <td>TTG</td>\n",
       "      <td>T</td>\n",
       "      <td>31.77</td>\n",
       "      <td>MG_INDEL_Filter</td>\n",
       "      <td>AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-1.391...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:30,5:35:60:60,0,838</td>\n",
       "      <td>...</td>\n",
       "      <td>QD=0.91</td>\n",
       "      <td>ReadPosRankSum=0.535</td>\n",
       "      <td>SOR=0.313</td>\n",
       "      <td>set=FilteredInAll</td>\n",
       "      <td>DEL</td>\n",
       "      <td>HET</td>\n",
       "      <td>VARTYPE=DEL</td>\n",
       "      <td>ANN=T|intron_variant|MODIFIER|Rb1cc1|gene-Rb1c...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6841143</td>\n",
       "      <td>.</td>\n",
       "      <td>CT</td>\n",
       "      <td>C</td>\n",
       "      <td>75.77</td>\n",
       "      <td>MG_INDEL_Filter</td>\n",
       "      <td>AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=0.727;...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:45,11:58:99:104,0,830</td>\n",
       "      <td>...</td>\n",
       "      <td>QD=1.35</td>\n",
       "      <td>ReadPosRankSum=0.815</td>\n",
       "      <td>SOR=0.725</td>\n",
       "      <td>set=FilteredInAll</td>\n",
       "      <td>DEL</td>\n",
       "      <td>HET</td>\n",
       "      <td>VARTYPE=DEL</td>\n",
       "      <td>ANN=C|intron_variant|MODIFIER|St18|gene-St18|t...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6872919</td>\n",
       "      <td>.</td>\n",
       "      <td>GC</td>\n",
       "      <td>G</td>\n",
       "      <td>4570.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=2;SAMPLES_AF=1.00;AN=2;BaseQRankSum=0.130;C...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>1/1:8,142:150:99:4599,183,0</td>\n",
       "      <td>...</td>\n",
       "      <td>QD=30.47</td>\n",
       "      <td>ReadPosRankSum=0.163</td>\n",
       "      <td>SOR=0.446</td>\n",
       "      <td>set=variant2</td>\n",
       "      <td>DEL</td>\n",
       "      <td>HOM</td>\n",
       "      <td>VARTYPE=DEL</td>\n",
       "      <td>ANN=G|frameshift_variant|HIGH|St18|gene-St18|t...</td>\n",
       "      <td>LOF=(St18|gene-St18|29|0.93)</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6878048</td>\n",
       "      <td>.</td>\n",
       "      <td>AT</td>\n",
       "      <td>A</td>\n",
       "      <td>53.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-3.400...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:11,5:16:82:82,0,224</td>\n",
       "      <td>...</td>\n",
       "      <td>QD=3.36</td>\n",
       "      <td>ReadPosRankSum=0.566</td>\n",
       "      <td>SOR=0.223</td>\n",
       "      <td>set=variant2</td>\n",
       "      <td>DEL</td>\n",
       "      <td>HET</td>\n",
       "      <td>VARTYPE=DEL</td>\n",
       "      <td>ANN=A|intron_variant|MODIFIER|St18|gene-St18|t...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10338</th>\n",
       "      <td>chrX</td>\n",
       "      <td>166090512</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>1729.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-5.350...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:97,73:170:99:1758,0,2595</td>\n",
       "      <td>...</td>\n",
       "      <td>QD=10.18</td>\n",
       "      <td>ReadPosRankSum=1.40</td>\n",
       "      <td>SOR=0.923</td>\n",
       "      <td>set=variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>HET</td>\n",
       "      <td>VARTYPE=SNP</td>\n",
       "      <td>ANN=T|missense_variant|MODERATE|Tlr7|gene-Tlr7...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10339</th>\n",
       "      <td>chrX</td>\n",
       "      <td>166260331</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>2804.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=1.91;C...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:75,103:178:99:2833,0,1717</td>\n",
       "      <td>...</td>\n",
       "      <td>QD=15.76</td>\n",
       "      <td>ReadPosRankSum=4.02</td>\n",
       "      <td>SOR=0.718</td>\n",
       "      <td>set=variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>HET</td>\n",
       "      <td>VARTYPE=SNP</td>\n",
       "      <td>ANN=A|missense_variant|MODERATE|Frmpd4|gene-Fr...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10340</th>\n",
       "      <td>chrX</td>\n",
       "      <td>167579241</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>604.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-1.100...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:40,29:69:99:633,0,1024</td>\n",
       "      <td>...</td>\n",
       "      <td>QD=8.76</td>\n",
       "      <td>ReadPosRankSum=2.50</td>\n",
       "      <td>SOR=0.604</td>\n",
       "      <td>set=variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>HET</td>\n",
       "      <td>VARTYPE=SNP</td>\n",
       "      <td>ANN=T|missense_variant|MODERATE|Arhgap6|gene-A...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10341</th>\n",
       "      <td>chrX</td>\n",
       "      <td>168084638</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>218.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=3.57;C...</td>\n",
       "      <td>GT:AD:DP:GQ:PGT:PID:PL</td>\n",
       "      <td>0/1:13,7:20:99:0|1:168084632_G_T:247,0,342</td>\n",
       "      <td>...</td>\n",
       "      <td>QD=10.94</td>\n",
       "      <td>ReadPosRankSum=-7.920e-01</td>\n",
       "      <td>SOR=1.402</td>\n",
       "      <td>set=variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>HET</td>\n",
       "      <td>VARTYPE=SNP</td>\n",
       "      <td>ANN=C|intron_variant|MODIFIER|Arhgap6|gene-Arh...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10342</th>\n",
       "      <td>chrX</td>\n",
       "      <td>168559972</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1803.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-1.983...</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:158,86:244:99:1832,0,4168</td>\n",
       "      <td>...</td>\n",
       "      <td>QD=7.39</td>\n",
       "      <td>ReadPosRankSum=2.53</td>\n",
       "      <td>SOR=0.529</td>\n",
       "      <td>set=variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>HET</td>\n",
       "      <td>VARTYPE=SNP</td>\n",
       "      <td>ANN=T|intron_variant|MODIFIER|Mid1|gene-Mid1|t...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10343 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      #CHROM        POS ID  REF ALT     QUAL           FILTER  \\\n",
       "0       chr1    3740866  .   TC   T  3487.77             PASS   \n",
       "1       chr1    6304093  .  TTG   T    31.77  MG_INDEL_Filter   \n",
       "2       chr1    6841143  .   CT   C    75.77  MG_INDEL_Filter   \n",
       "3       chr1    6872919  .   GC   G  4570.77             PASS   \n",
       "4       chr1    6878048  .   AT   A    53.77             PASS   \n",
       "...      ...        ... ..  ...  ..      ...              ...   \n",
       "10338   chrX  166090512  .    A   T  1729.77             PASS   \n",
       "10339   chrX  166260331  .    T   A  2804.77             PASS   \n",
       "10340   chrX  167579241  .    A   T   604.77             PASS   \n",
       "10341   chrX  168084638  .    T   C   218.77             PASS   \n",
       "10342   chrX  168559972  .    G   T  1803.77             PASS   \n",
       "\n",
       "                                                    INFO  \\\n",
       "0      AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-3.460...   \n",
       "1      AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-1.391...   \n",
       "2      AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=0.727;...   \n",
       "3      AC=2;SAMPLES_AF=1.00;AN=2;BaseQRankSum=0.130;C...   \n",
       "4      AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-3.400...   \n",
       "...                                                  ...   \n",
       "10338  AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-5.350...   \n",
       "10339  AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=1.91;C...   \n",
       "10340  AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-1.100...   \n",
       "10341  AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=3.57;C...   \n",
       "10342  AC=1;SAMPLES_AF=0.500;AN=2;BaseQRankSum=-1.983...   \n",
       "\n",
       "                       FORMAT                                       100M1  \\\n",
       "0              GT:AD:DP:GQ:PL                 0/1:10,104:114:30:3516,0,30   \n",
       "1              GT:AD:DP:GQ:PL                     0/1:30,5:35:60:60,0,838   \n",
       "2              GT:AD:DP:GQ:PL                   0/1:45,11:58:99:104,0,830   \n",
       "3              GT:AD:DP:GQ:PL                 1/1:8,142:150:99:4599,183,0   \n",
       "4              GT:AD:DP:GQ:PL                     0/1:11,5:16:82:82,0,224   \n",
       "...                       ...                                         ...   \n",
       "10338          GT:AD:DP:GQ:PL                0/1:97,73:170:99:1758,0,2595   \n",
       "10339          GT:AD:DP:GQ:PL               0/1:75,103:178:99:2833,0,1717   \n",
       "10340          GT:AD:DP:GQ:PL                  0/1:40,29:69:99:633,0,1024   \n",
       "10341  GT:AD:DP:GQ:PGT:PID:PL  0/1:13,7:20:99:0|1:168084632_G_T:247,0,342   \n",
       "10342          GT:AD:DP:GQ:PL               0/1:158,86:244:99:1832,0,4168   \n",
       "\n",
       "       ...   INFO_12                    INFO_13    INFO_14            INFO_15  \\\n",
       "0      ...  QD=30.59       ReadPosRankSum=0.175  SOR=0.625       set=variant2   \n",
       "1      ...   QD=0.91       ReadPosRankSum=0.535  SOR=0.313  set=FilteredInAll   \n",
       "2      ...   QD=1.35       ReadPosRankSum=0.815  SOR=0.725  set=FilteredInAll   \n",
       "3      ...  QD=30.47       ReadPosRankSum=0.163  SOR=0.446       set=variant2   \n",
       "4      ...   QD=3.36       ReadPosRankSum=0.566  SOR=0.223       set=variant2   \n",
       "...    ...       ...                        ...        ...                ...   \n",
       "10338  ...  QD=10.18        ReadPosRankSum=1.40  SOR=0.923        set=variant   \n",
       "10339  ...  QD=15.76        ReadPosRankSum=4.02  SOR=0.718        set=variant   \n",
       "10340  ...   QD=8.76        ReadPosRankSum=2.50  SOR=0.604        set=variant   \n",
       "10341  ...  QD=10.94  ReadPosRankSum=-7.920e-01  SOR=1.402        set=variant   \n",
       "10342  ...   QD=7.39        ReadPosRankSum=2.53  SOR=0.529        set=variant   \n",
       "\n",
       "      INFO_16 INFO_17      INFO_18  \\\n",
       "0         DEL     HET  VARTYPE=DEL   \n",
       "1         DEL     HET  VARTYPE=DEL   \n",
       "2         DEL     HET  VARTYPE=DEL   \n",
       "3         DEL     HOM  VARTYPE=DEL   \n",
       "4         DEL     HET  VARTYPE=DEL   \n",
       "...       ...     ...          ...   \n",
       "10338     SNP     HET  VARTYPE=SNP   \n",
       "10339     SNP     HET  VARTYPE=SNP   \n",
       "10340     SNP     HET  VARTYPE=SNP   \n",
       "10341     SNP     HET  VARTYPE=SNP   \n",
       "10342     SNP     HET  VARTYPE=SNP   \n",
       "\n",
       "                                                 INFO_19  \\\n",
       "0      ANN=T|frameshift_variant|HIGH|Xkr4|gene-Xkr4|t...   \n",
       "1      ANN=T|intron_variant|MODIFIER|Rb1cc1|gene-Rb1c...   \n",
       "2      ANN=C|intron_variant|MODIFIER|St18|gene-St18|t...   \n",
       "3      ANN=G|frameshift_variant|HIGH|St18|gene-St18|t...   \n",
       "4      ANN=A|intron_variant|MODIFIER|St18|gene-St18|t...   \n",
       "...                                                  ...   \n",
       "10338  ANN=T|missense_variant|MODERATE|Tlr7|gene-Tlr7...   \n",
       "10339  ANN=A|missense_variant|MODERATE|Frmpd4|gene-Fr...   \n",
       "10340  ANN=T|missense_variant|MODERATE|Arhgap6|gene-A...   \n",
       "10341  ANN=C|intron_variant|MODIFIER|Arhgap6|gene-Arh...   \n",
       "10342  ANN=T|intron_variant|MODIFIER|Mid1|gene-Mid1|t...   \n",
       "\n",
       "                            INFO_20 INFO_21  \n",
       "0       LOF=(Xkr4|gene-Xkr4|4|1.00)    None  \n",
       "1                              None    None  \n",
       "2                              None    None  \n",
       "3      LOF=(St18|gene-St18|29|0.93)    None  \n",
       "4                              None    None  \n",
       "...                             ...     ...  \n",
       "10338                          None    None  \n",
       "10339                          None    None  \n",
       "10340                          None    None  \n",
       "10341                          None    None  \n",
       "10342                          None    None  \n",
       "\n",
       "[10343 rows x 31 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "e767e338",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-10T23:54:36.820710Z",
     "start_time": "2023-11-10T23:54:36.811349Z"
    }
   },
   "outputs": [],
   "source": [
    "df = df.drop('INFO', axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "6e3ef4e4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-10T23:54:37.431884Z",
     "start_time": "2023-11-10T23:54:37.403069Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>100M1</th>\n",
       "      <th>INFO_1</th>\n",
       "      <th>...</th>\n",
       "      <th>INFO_12</th>\n",
       "      <th>INFO_13</th>\n",
       "      <th>INFO_14</th>\n",
       "      <th>INFO_15</th>\n",
       "      <th>INFO_16</th>\n",
       "      <th>INFO_17</th>\n",
       "      <th>INFO_18</th>\n",
       "      <th>INFO_19</th>\n",
       "      <th>INFO_20</th>\n",
       "      <th>INFO_21</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3740866</td>\n",
       "      <td>.</td>\n",
       "      <td>TC</td>\n",
       "      <td>T</td>\n",
       "      <td>3487.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:10,104:114:30:3516,0,30</td>\n",
       "      <td>AC=1</td>\n",
       "      <td>...</td>\n",
       "      <td>QD=30.59</td>\n",
       "      <td>ReadPosRankSum=0.175</td>\n",
       "      <td>SOR=0.625</td>\n",
       "      <td>set=variant2</td>\n",
       "      <td>DEL</td>\n",
       "      <td>HET</td>\n",
       "      <td>VARTYPE=DEL</td>\n",
       "      <td>ANN=T|frameshift_variant|HIGH|Xkr4|gene-Xkr4|t...</td>\n",
       "      <td>LOF=(Xkr4|gene-Xkr4|4|1.00)</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6304093</td>\n",
       "      <td>.</td>\n",
       "      <td>TTG</td>\n",
       "      <td>T</td>\n",
       "      <td>31.77</td>\n",
       "      <td>MG_INDEL_Filter</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:30,5:35:60:60,0,838</td>\n",
       "      <td>AC=1</td>\n",
       "      <td>...</td>\n",
       "      <td>QD=0.91</td>\n",
       "      <td>ReadPosRankSum=0.535</td>\n",
       "      <td>SOR=0.313</td>\n",
       "      <td>set=FilteredInAll</td>\n",
       "      <td>DEL</td>\n",
       "      <td>HET</td>\n",
       "      <td>VARTYPE=DEL</td>\n",
       "      <td>ANN=T|intron_variant|MODIFIER|Rb1cc1|gene-Rb1c...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6841143</td>\n",
       "      <td>.</td>\n",
       "      <td>CT</td>\n",
       "      <td>C</td>\n",
       "      <td>75.77</td>\n",
       "      <td>MG_INDEL_Filter</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:45,11:58:99:104,0,830</td>\n",
       "      <td>AC=1</td>\n",
       "      <td>...</td>\n",
       "      <td>QD=1.35</td>\n",
       "      <td>ReadPosRankSum=0.815</td>\n",
       "      <td>SOR=0.725</td>\n",
       "      <td>set=FilteredInAll</td>\n",
       "      <td>DEL</td>\n",
       "      <td>HET</td>\n",
       "      <td>VARTYPE=DEL</td>\n",
       "      <td>ANN=C|intron_variant|MODIFIER|St18|gene-St18|t...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6872919</td>\n",
       "      <td>.</td>\n",
       "      <td>GC</td>\n",
       "      <td>G</td>\n",
       "      <td>4570.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>1/1:8,142:150:99:4599,183,0</td>\n",
       "      <td>AC=2</td>\n",
       "      <td>...</td>\n",
       "      <td>QD=30.47</td>\n",
       "      <td>ReadPosRankSum=0.163</td>\n",
       "      <td>SOR=0.446</td>\n",
       "      <td>set=variant2</td>\n",
       "      <td>DEL</td>\n",
       "      <td>HOM</td>\n",
       "      <td>VARTYPE=DEL</td>\n",
       "      <td>ANN=G|frameshift_variant|HIGH|St18|gene-St18|t...</td>\n",
       "      <td>LOF=(St18|gene-St18|29|0.93)</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6878048</td>\n",
       "      <td>.</td>\n",
       "      <td>AT</td>\n",
       "      <td>A</td>\n",
       "      <td>53.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:11,5:16:82:82,0,224</td>\n",
       "      <td>AC=1</td>\n",
       "      <td>...</td>\n",
       "      <td>QD=3.36</td>\n",
       "      <td>ReadPosRankSum=0.566</td>\n",
       "      <td>SOR=0.223</td>\n",
       "      <td>set=variant2</td>\n",
       "      <td>DEL</td>\n",
       "      <td>HET</td>\n",
       "      <td>VARTYPE=DEL</td>\n",
       "      <td>ANN=A|intron_variant|MODIFIER|St18|gene-St18|t...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10338</th>\n",
       "      <td>chrX</td>\n",
       "      <td>166090512</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>1729.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:97,73:170:99:1758,0,2595</td>\n",
       "      <td>AC=1</td>\n",
       "      <td>...</td>\n",
       "      <td>QD=10.18</td>\n",
       "      <td>ReadPosRankSum=1.40</td>\n",
       "      <td>SOR=0.923</td>\n",
       "      <td>set=variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>HET</td>\n",
       "      <td>VARTYPE=SNP</td>\n",
       "      <td>ANN=T|missense_variant|MODERATE|Tlr7|gene-Tlr7...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10339</th>\n",
       "      <td>chrX</td>\n",
       "      <td>166260331</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>2804.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:75,103:178:99:2833,0,1717</td>\n",
       "      <td>AC=1</td>\n",
       "      <td>...</td>\n",
       "      <td>QD=15.76</td>\n",
       "      <td>ReadPosRankSum=4.02</td>\n",
       "      <td>SOR=0.718</td>\n",
       "      <td>set=variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>HET</td>\n",
       "      <td>VARTYPE=SNP</td>\n",
       "      <td>ANN=A|missense_variant|MODERATE|Frmpd4|gene-Fr...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10340</th>\n",
       "      <td>chrX</td>\n",
       "      <td>167579241</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>604.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:40,29:69:99:633,0,1024</td>\n",
       "      <td>AC=1</td>\n",
       "      <td>...</td>\n",
       "      <td>QD=8.76</td>\n",
       "      <td>ReadPosRankSum=2.50</td>\n",
       "      <td>SOR=0.604</td>\n",
       "      <td>set=variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>HET</td>\n",
       "      <td>VARTYPE=SNP</td>\n",
       "      <td>ANN=T|missense_variant|MODERATE|Arhgap6|gene-A...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10341</th>\n",
       "      <td>chrX</td>\n",
       "      <td>168084638</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>218.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>GT:AD:DP:GQ:PGT:PID:PL</td>\n",
       "      <td>0/1:13,7:20:99:0|1:168084632_G_T:247,0,342</td>\n",
       "      <td>AC=1</td>\n",
       "      <td>...</td>\n",
       "      <td>QD=10.94</td>\n",
       "      <td>ReadPosRankSum=-7.920e-01</td>\n",
       "      <td>SOR=1.402</td>\n",
       "      <td>set=variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>HET</td>\n",
       "      <td>VARTYPE=SNP</td>\n",
       "      <td>ANN=C|intron_variant|MODIFIER|Arhgap6|gene-Arh...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10342</th>\n",
       "      <td>chrX</td>\n",
       "      <td>168559972</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1803.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0/1:158,86:244:99:1832,0,4168</td>\n",
       "      <td>AC=1</td>\n",
       "      <td>...</td>\n",
       "      <td>QD=7.39</td>\n",
       "      <td>ReadPosRankSum=2.53</td>\n",
       "      <td>SOR=0.529</td>\n",
       "      <td>set=variant</td>\n",
       "      <td>SNP</td>\n",
       "      <td>HET</td>\n",
       "      <td>VARTYPE=SNP</td>\n",
       "      <td>ANN=T|intron_variant|MODIFIER|Mid1|gene-Mid1|t...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10343 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      #CHROM        POS ID  REF ALT     QUAL           FILTER  \\\n",
       "0       chr1    3740866  .   TC   T  3487.77             PASS   \n",
       "1       chr1    6304093  .  TTG   T    31.77  MG_INDEL_Filter   \n",
       "2       chr1    6841143  .   CT   C    75.77  MG_INDEL_Filter   \n",
       "3       chr1    6872919  .   GC   G  4570.77             PASS   \n",
       "4       chr1    6878048  .   AT   A    53.77             PASS   \n",
       "...      ...        ... ..  ...  ..      ...              ...   \n",
       "10338   chrX  166090512  .    A   T  1729.77             PASS   \n",
       "10339   chrX  166260331  .    T   A  2804.77             PASS   \n",
       "10340   chrX  167579241  .    A   T   604.77             PASS   \n",
       "10341   chrX  168084638  .    T   C   218.77             PASS   \n",
       "10342   chrX  168559972  .    G   T  1803.77             PASS   \n",
       "\n",
       "                       FORMAT                                       100M1  \\\n",
       "0              GT:AD:DP:GQ:PL                 0/1:10,104:114:30:3516,0,30   \n",
       "1              GT:AD:DP:GQ:PL                     0/1:30,5:35:60:60,0,838   \n",
       "2              GT:AD:DP:GQ:PL                   0/1:45,11:58:99:104,0,830   \n",
       "3              GT:AD:DP:GQ:PL                 1/1:8,142:150:99:4599,183,0   \n",
       "4              GT:AD:DP:GQ:PL                     0/1:11,5:16:82:82,0,224   \n",
       "...                       ...                                         ...   \n",
       "10338          GT:AD:DP:GQ:PL                0/1:97,73:170:99:1758,0,2595   \n",
       "10339          GT:AD:DP:GQ:PL               0/1:75,103:178:99:2833,0,1717   \n",
       "10340          GT:AD:DP:GQ:PL                  0/1:40,29:69:99:633,0,1024   \n",
       "10341  GT:AD:DP:GQ:PGT:PID:PL  0/1:13,7:20:99:0|1:168084632_G_T:247,0,342   \n",
       "10342          GT:AD:DP:GQ:PL               0/1:158,86:244:99:1832,0,4168   \n",
       "\n",
       "      INFO_1  ...   INFO_12                    INFO_13    INFO_14  \\\n",
       "0       AC=1  ...  QD=30.59       ReadPosRankSum=0.175  SOR=0.625   \n",
       "1       AC=1  ...   QD=0.91       ReadPosRankSum=0.535  SOR=0.313   \n",
       "2       AC=1  ...   QD=1.35       ReadPosRankSum=0.815  SOR=0.725   \n",
       "3       AC=2  ...  QD=30.47       ReadPosRankSum=0.163  SOR=0.446   \n",
       "4       AC=1  ...   QD=3.36       ReadPosRankSum=0.566  SOR=0.223   \n",
       "...      ...  ...       ...                        ...        ...   \n",
       "10338   AC=1  ...  QD=10.18        ReadPosRankSum=1.40  SOR=0.923   \n",
       "10339   AC=1  ...  QD=15.76        ReadPosRankSum=4.02  SOR=0.718   \n",
       "10340   AC=1  ...   QD=8.76        ReadPosRankSum=2.50  SOR=0.604   \n",
       "10341   AC=1  ...  QD=10.94  ReadPosRankSum=-7.920e-01  SOR=1.402   \n",
       "10342   AC=1  ...   QD=7.39        ReadPosRankSum=2.53  SOR=0.529   \n",
       "\n",
       "                 INFO_15 INFO_16 INFO_17      INFO_18  \\\n",
       "0           set=variant2     DEL     HET  VARTYPE=DEL   \n",
       "1      set=FilteredInAll     DEL     HET  VARTYPE=DEL   \n",
       "2      set=FilteredInAll     DEL     HET  VARTYPE=DEL   \n",
       "3           set=variant2     DEL     HOM  VARTYPE=DEL   \n",
       "4           set=variant2     DEL     HET  VARTYPE=DEL   \n",
       "...                  ...     ...     ...          ...   \n",
       "10338        set=variant     SNP     HET  VARTYPE=SNP   \n",
       "10339        set=variant     SNP     HET  VARTYPE=SNP   \n",
       "10340        set=variant     SNP     HET  VARTYPE=SNP   \n",
       "10341        set=variant     SNP     HET  VARTYPE=SNP   \n",
       "10342        set=variant     SNP     HET  VARTYPE=SNP   \n",
       "\n",
       "                                                 INFO_19  \\\n",
       "0      ANN=T|frameshift_variant|HIGH|Xkr4|gene-Xkr4|t...   \n",
       "1      ANN=T|intron_variant|MODIFIER|Rb1cc1|gene-Rb1c...   \n",
       "2      ANN=C|intron_variant|MODIFIER|St18|gene-St18|t...   \n",
       "3      ANN=G|frameshift_variant|HIGH|St18|gene-St18|t...   \n",
       "4      ANN=A|intron_variant|MODIFIER|St18|gene-St18|t...   \n",
       "...                                                  ...   \n",
       "10338  ANN=T|missense_variant|MODERATE|Tlr7|gene-Tlr7...   \n",
       "10339  ANN=A|missense_variant|MODERATE|Frmpd4|gene-Fr...   \n",
       "10340  ANN=T|missense_variant|MODERATE|Arhgap6|gene-A...   \n",
       "10341  ANN=C|intron_variant|MODIFIER|Arhgap6|gene-Arh...   \n",
       "10342  ANN=T|intron_variant|MODIFIER|Mid1|gene-Mid1|t...   \n",
       "\n",
       "                            INFO_20 INFO_21  \n",
       "0       LOF=(Xkr4|gene-Xkr4|4|1.00)    None  \n",
       "1                              None    None  \n",
       "2                              None    None  \n",
       "3      LOF=(St18|gene-St18|29|0.93)    None  \n",
       "4                              None    None  \n",
       "...                             ...     ...  \n",
       "10338                          None    None  \n",
       "10339                          None    None  \n",
       "10340                          None    None  \n",
       "10341                          None    None  \n",
       "10342                          None    None  \n",
       "\n",
       "[10343 rows x 30 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13084534",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "datascience",
   "language": "python",
   "name": "datascience"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
